File: BusinessSDK/Sources/BusinessSDKCore/NetworkCore.swift
import Foundation

public enum HTTPMethod: String, Sendable {
    case get = "GET"
    case post = "POST"
    case put = "PUT"
    case delete = "DELETE"
}

public protocol APIRequest: Sendable {
    associatedtype Response: Decodable & Sendable

    var path: String { get }
    var method: HTTPMethod { get }
    var body: Data? { get }
    var headers: [String: String] { get }
    var queryItems: [URLQueryItem] { get }
    var requiresAuthorization: Bool { get }
}

public enum NetworkError: Error, Sendable {
    case url(URLError)
    case invalidResponse
    case unauthorized
    case httpStatus(code: Int, data: Data)
    case noData
    case decoding(Error)
}

public protocol SecurityKitProtocol: Sendable {
    func getAccessToken() async -> String?
    func refreshAccessToken() async throws
    func clearTokens() async
}

public actor NoOpSecurityKit: SecurityKitProtocol {
    public init() {}

    public func getAccessToken() async -> String? { nil }
    public func refreshAccessToken() async throws {}
    public func clearTokens() async {}
}

public actor RefreshGate {
    public init() {}

    public func refreshIfNeeded(_ operation: @Sendable () async throws -> Void) async throws {
        try await operation()
    }
}

public protocol NetworkEngineProtocol: Sendable {
    func request<T: APIRequest>(_ requestConfig: T) async throws -> T.Response
}

public final class NetworkEngine: NetworkEngineProtocol, @unchecked Sendable {
    private let baseURL: URL
    private let session: URLSession
    private let security: SecurityKitProtocol
    private let decoder: JSONDecoder
    private let refreshGate: RefreshGate

    public init(
        baseURL: URL,
        security: SecurityKitProtocol = NoOpSecurityKit(),
        session: URLSession = .shared,
        decoder: JSONDecoder = .init(),
        refreshGate: RefreshGate = RefreshGate()
    ) {
        self.baseURL = baseURL
        self.security = security
        self.session = session
        self.decoder = decoder
        self.refreshGate = refreshGate
    }

    public func request<T: APIRequest>(_ requestConfig: T) async throws -> T.Response {
        let url = try makeURL(for: requestConfig)
        var request = URLRequest(url: url)
        request.httpMethod = requestConfig.method.rawValue
        request.httpBody = requestConfig.body

        requestConfig.headers.forEach { request.setValue($0.value, forHTTPHeaderField: $0.key) }

        if requestConfig.requiresAuthorization, let token = await security.getAccessToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        return try await perform(request, as: T.self, didRetryAfterRefresh: false)
    }

    private func makeURL<T: APIRequest>(for requestConfig: T) throws -> URL {
        let endpoint = requestConfig.path.hasPrefix("/") ? String(requestConfig.path.dropFirst()) : requestConfig.path
        let url = baseURL.appendingPathComponent(endpoint)

        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            throw NetworkError.invalidResponse
        }

        if !requestConfig.queryItems.isEmpty {
            components.queryItems = requestConfig.queryItems
        }

        guard let finalURL = components.url else {
            throw NetworkError.invalidResponse
        }

        return finalURL
    }

    private func perform<T: APIRequest>(
        _ request: URLRequest,
        as _: T.Type,
        didRetryAfterRefresh: Bool
    ) async throws -> T.Response {
        let data: Data
        let response: URLResponse

        do {
            (data, response) = try await session.data(for: request)
        } catch let error as URLError {
            throw NetworkError.url(error)
        } catch {
            throw error
        }

        guard let http = response as? HTTPURLResponse else {
            throw NetworkError.invalidResponse
        }

        if http.statusCode == 401 {
            if didRetryAfterRefresh {
                await security.clearTokens()
                throw NetworkError.unauthorized
            }

            let hadAuthHeader = request.value(forHTTPHeaderField: "Authorization") != nil
            guard hadAuthHeader else {
                throw NetworkError.unauthorized
            }

            do {
                try await refreshGate.refreshIfNeeded {
                    try await self.security.refreshAccessToken()
                }
            } catch {
                await security.clearTokens()
                throw NetworkError.unauthorized
            }

            guard let newToken = await security.getAccessToken() else {
                await security.clearTokens()
                throw NetworkError.unauthorized
            }

            var retry = request
            retry.setValue("Bearer \(newToken)", forHTTPHeaderField: "Authorization")
            return try await perform(retry, as: T.self, didRetryAfterRefresh: true)
        }

        guard (200...299).contains(http.statusCode) else {
            throw NetworkError.httpStatus(code: http.statusCode, data: data)
        }

        guard !data.isEmpty else {
            throw NetworkError.noData
        }

        do {
            return try decoder.decode(T.Response.self, from: data)
        } catch {
            throw NetworkError.decoding(error)
        }
    }
}

File: android/sdk-core/src/main/java/com/multiplatformbusinesssdk/core/NetworkCore.kt
package com.multiplatformbusinesssdk.core

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

enum class HttpMethod {
    GET,
    POST,
    PUT,
    DELETE
}

sealed class NetworkError(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class Url(cause: IOException) : NetworkError("URL error", cause)
    object InvalidResponse : NetworkError("Invalid HTTP response")
    object Unauthorized : NetworkError("Unauthorized")
    class HttpStatus(val code: Int, val payload: String) : NetworkError("Unexpected status: $code")
    object NoData : NetworkError("No data returned")
    class Decoding(cause: Throwable) : NetworkError("Decoding failed", cause)
}

interface TokenStore {
    suspend fun getAccessToken(): String?
    suspend fun refreshAccessToken()
    suspend fun clearTokens()
}

object NoOpTokenStore : TokenStore {
    override suspend fun getAccessToken(): String? = null
    override suspend fun refreshAccessToken() = Unit
    override suspend fun clearTokens() = Unit
}

class RefreshGate {
    private val mutex = Mutex()

    suspend fun refreshIfNeeded(operation: suspend () -> Unit) {
        mutex.withLock {
            operation()
        }
    }
}

interface ApiRequest<T> {
    val path: String
    val method: HttpMethod
    val body: ByteArray?
    val headers: Map<String, String>
    val query: Map<String, String>
    val requiresAuthorization: Boolean

    fun parse(payload: String): T
}

class NetworkEngine(
    private val baseUrl: String,
    private val tokenStore: TokenStore = NoOpTokenStore,
    private val refreshGate: RefreshGate = RefreshGate()
) {
    suspend fun <T> request(requestConfig: ApiRequest<T>): T {
        val url = buildUrl(requestConfig)
        return perform(url, requestConfig, didRetryAfterRefresh = false)
    }

    private suspend fun <T> perform(
        url: URL,
        requestConfig: ApiRequest<T>,
        didRetryAfterRefresh: Boolean
    ): T {
        val connection = (url.openConnection() as? HttpURLConnection) ?: throw NetworkError.InvalidResponse

        try {
            connection.requestMethod = requestConfig.method.name
            connection.instanceFollowRedirects = true
            requestConfig.headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }

            if (requestConfig.requiresAuthorization) {
                tokenStore.getAccessToken()?.let { token ->
                    connection.setRequestProperty("Authorization", "Bearer $token")
                }
            }

            requestConfig.body?.let { body ->
                connection.doOutput = true
                connection.outputStream.use { stream -> stream.write(body) }
            }

            val statusCode = connection.responseCode
            val payload = readPayload(connection, statusCode)

            if (statusCode == 401) {
                if (didRetryAfterRefresh) {
                    tokenStore.clearTokens()
                    throw NetworkError.Unauthorized
                }

                val hadAuthHeader = connection.getRequestProperty("Authorization") != null
                if (!hadAuthHeader) {
                    throw NetworkError.Unauthorized
                }

                try {
                    refreshGate.refreshIfNeeded {
                        tokenStore.refreshAccessToken()
                    }
                } catch (_: Throwable) {
                    tokenStore.clearTokens()
                    throw NetworkError.Unauthorized
                }

                val refreshedToken = tokenStore.getAccessToken() ?: run {
                    tokenStore.clearTokens()
                    throw NetworkError.Unauthorized
                }

                val retryHeaders = requestConfig.headers + mapOf("Authorization" to "Bearer $refreshedToken")
                val retryRequest = object : ApiRequest<T> by requestConfig {
                    override val headers: Map<String, String> = retryHeaders
                }
                return perform(buildUrl(retryRequest), retryRequest, didRetryAfterRefresh = true)
            }

            if (statusCode !in 200..299) {
                throw NetworkError.HttpStatus(statusCode, payload)
            }

            if (payload.isBlank()) {
                throw NetworkError.NoData
            }

            return try {
                requestConfig.parse(payload)
            } catch (error: Throwable) {
                throw NetworkError.Decoding(error)
            }
        } catch (error: IOException) {
            throw NetworkError.Url(error)
        } finally {
            connection.disconnect()
        }
    }

    private fun <T> buildUrl(requestConfig: ApiRequest<T>): URL {
        val normalizedBaseUrl = baseUrl.trimEnd('/')
        val normalizedPath = requestConfig.path.trimStart('/')
        val query = if (requestConfig.query.isEmpty()) {
            ""
        } else {
            requestConfig.query.entries.joinToString(prefix = "?", separator = "&") { (key, value) ->
                "${encode(key)}=${encode(value)}"
            }
        }
        return URL("$normalizedBaseUrl/$normalizedPath$query")
    }

    private fun encode(value: String): String = java.net.URLEncoder.encode(value, Charsets.UTF_8.name())

    private fun readPayload(connection: HttpURLConnection, statusCode: Int): String {
        val stream = if (statusCode in 200..299) connection.inputStream else connection.errorStream
        return stream?.bufferedReader()?.use { it.readText() }.orEmpty()
    }
}
